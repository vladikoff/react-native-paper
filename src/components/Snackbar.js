/* @flow */

import * as React from 'react';
import {
  StyleSheet,
  Animated,
  Text,
  View,
  TouchableWithoutFeedback,
} from 'react-native';

import ThemedPortal from './Portal/ThemedPortal';
import withTheme from '../core/withTheme';
import { white } from '../styles/colors';
import type { Theme } from '../types';

type Props = {
  /**
   * Whether snackbar is currently visible.
   */
  visible: boolean,
  /**
   * Text that will be displayed inside SnackBar.
   */
  children: React.Node,
  /**
   * Object that determines button text and callback for button press. It should contains following properties:
   * - `text` - Content of the action button
   * - `onPress` - Callback that is called when action button is pressed, user needs to update the `visible` prop.
   */
  action?: {
    text: string,
    onPress: () => mixed,
  },
  /**
   * The duration for which the Snackbar is shown
   * It can take following values:
   * - `Snackbar.DURATION_INDEFINITE` - SnackBar will hide only when user tap it.
   * - `Snackbar.DURATION_SHORT` - SnackBar will hide after 2 seconds.
   * - `Snackbar.DURATION_LONG` - SnackBar will hide after 3.5 seconds.
   */
  duration?: number,
  /**
   * Callback called when Snackbar is dismissed. The `visible` prop needs to be updated when this is called.
   */
  onDismiss: () => mixed,
  style?: any,
  theme: Theme,
};

type State = {
  rendered: boolean,
  height: number,
  opacity: Animated.Value,
  yPosition: Animated.Value,
};

const SNACKBAR_ANIMATION_DURATION = 250;

/**
 * Snackbar provide brief feedback about an operation through a message at the bottom of the screen.
 *
 * <div class="screenshots">
 *   <img class="medium" src="screenshots/snackbar.gif" />
 * </div>
 *
 * ## Usage
 * ```js
 * import React from 'react';
 * import { Snackbar, StyleSheet } from 'react-native-paper';
 *
 * export default class MyComponent extends React.Component {
 *   state = {
 *     visible: false,
 *   };
 *
 *   render() {
 *     const { visible } = this.state;
 *     return (
 *       <View style={styles.container}>
 *         <View>
 *           <Button onPress={() => this.setState({ visible: true })}>Show</Button>
 *           <Button onPress={() => this.setState({ visible: false })}>Hide</Button>
 *         </View>
 *         <Snackbar
 *           visible={this.state.visible}
 *           onDismiss={() => this.setState({ visible: false })}
 *           action={{
 *             text: 'Undo',
 *             onPress: () => {
 *               this.setState({ visible: false });
 *             },
 *           }}
 *         >
 *           Put your message here
 *         </Snackbar>
 *       </View>
 *     );
 *   }
 * }
 *
 * const styles = StyleSheet.create({
 *   container: {
 *     flex: 1,
 *     justifyContent: 'space-between',
 *   },
 * });
 * ```
 */
class Snackbar extends React.Component<Props, State> {
  static defaultProps = {
    duration: this.DURATION_LONG,
  };

  state = {
    rendered: false,
    height: 0,
    opacity: new Animated.Value(0),
    yPosition: new Animated.Value(0),
  };

  _hideTimeout: any;

  static DURATION_SHORT = 2000;
  static DURATION_LONG = 3500;
  static DURATION_INDEFINITE = 0;

  componentDidMount() {
    if (this.props.visible) {
      this._show();
    }
  }

  componentDidUpdate(prevProps) {
    if (prevProps.visible !== this.props.visible) {
      this.props.visible ? this._show() : this._hide();
    }
  }

  componentWillUnmount() {
    if (this._hideTimeout) {
      clearTimeout(this._hideTimeout);
    }
  }

  _onLayout = ({ nativeEvent }) => {
    const {
      layout: { height },
    } = nativeEvent;

    this.setState({
      height,
      rendered: true,
    });

    this.state.yPosition.setValue(height);
  };

  _show = () => {
    Animated.parallel([
      Animated.timing(this.state.opacity, {
        toValue: 1,
        duration: SNACKBAR_ANIMATION_DURATION,
        useNativeDriver: true,
      }),
      Animated.timing(this.state.yPosition, {
        toValue: 0,
        duration: SNACKBAR_ANIMATION_DURATION,
        useNativeDriver: true,
      }),
    ]).start(() => {
      const { duration } = this.props;
      if (duration !== Snackbar.DURATION_INDEFINITE) {
        this._hideTimeout = setTimeout(this.props.onDismiss, duration);
      }
    });
  };

  _hide = () => {
    if (this._hideTimeout) {
      clearTimeout(this._hideTimeout);
    }
    Animated.parallel([
      Animated.timing(this.state.opacity, {
        toValue: 0,
        duration: SNACKBAR_ANIMATION_DURATION,
        useNativeDriver: true,
      }),
      Animated.timing(this.state.yPosition, {
        toValue: this.state.height,
        duration: SNACKBAR_ANIMATION_DURATION,
        useNativeDriver: true,
      }),
    ]).start();
  };

  render() {
    const { children, action, onDismiss, theme, style } = this.props;

    const { fonts, colors } = theme;

    const buttonMargin = action ? 24 : 0;
    const contentRightMargin = action ? 0 : 24;

    return (
      <ThemedPortal>
        <Animated.View
          onLayout={this._onLayout}
          style={[
            styles.wrapper,
            {
              opacity: this.state.rendered ? 1 : 0,
              transform: [
                {
                  translateY: this.state.yPosition,
                },
              ],
            },
            style,
          ]}
        >
          <TouchableWithoutFeedback onPress={onDismiss}>
            <Animated.View
              style={[
                styles.container,
                {
                  opacity: this.state.opacity.interpolate({
                    inputRange: [0, 0.8, 1],
                    outputRange: [0, 0.2, 1],
                  }),
                },
              ]}
            >
              <Text
                style={[
                  styles.content,
                  {
                    fontFamily: fonts.regular,
                    marginRight: contentRightMargin,
                  },
                ]}
              >
                {children}
              </Text>
              {action ? (
                <View
                  style={{
                    marginHorizontal: buttonMargin,
                  }}
                >
                  <TouchableWithoutFeedback
                    onPress={() => {
                      action.onPress();
                      this._hide();
                    }}
                  >
                    <View>
                      <Text style={{ color: colors.accent }}>
                        {action.text.toUpperCase()}
                      </Text>
                    </View>
                  </TouchableWithoutFeedback>
                </View>
              ) : null}
            </Animated.View>
          </TouchableWithoutFeedback>
        </Animated.View>
      </ThemedPortal>
    );
  }
}

const styles = StyleSheet.create({
  wrapper: {
    backgroundColor: '#323232',
    position: 'absolute',
    bottom: 0,
    width: '100%',
  },
  container: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  content: {
    color: white,
    marginLeft: 24,
    marginVertical: 14,
    flexWrap: 'wrap',
    flex: 1,
  },
});

export default withTheme(Snackbar);
