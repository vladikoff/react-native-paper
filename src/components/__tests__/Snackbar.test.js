/* @flow */

import * as React from 'react';
import renderer from 'react-test-renderer';
import { Text } from 'react-native';
import Snackbar from '../Snackbar';
import PortalHost from '../Portal/PortalHost';

jest.useFakeTimers();

const func = () => {};

it('renders snackbar with content', () => {
  const tree = renderer
    .create(
      <PortalHost>
        <Snackbar visible onDismiss={func}>
          Snackbar content
        </Snackbar>
      </PortalHost>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders not visible snackbar with content', () => {
  const tree = renderer
    .create(
      <PortalHost>
        <Snackbar visible={false} onDismiss={func}>
          Snackbar content
        </Snackbar>
      </PortalHost>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders snackbar with Text as a child', () => {
  const tree = renderer
    .create(
      <PortalHost>
        <Snackbar visible onDismiss={func}>
          <Text>Snackbar content</Text>
        </Snackbar>
      </PortalHost>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders snackbar with action button', () => {
  const tree = renderer
    .create(
      <PortalHost>
        <Snackbar
          visible
          onDismiss={() => {}}
          action={{ text: 'Undo', onPress: func }}
        >
          Snackbar content
        </Snackbar>
      </PortalHost>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
