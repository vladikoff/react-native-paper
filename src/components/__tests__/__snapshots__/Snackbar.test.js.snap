// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders not visible snackbar with content 1`] = `
<View
  pointerEvents="box-none"
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    pointerEvents="box-none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      collapsable={undefined}
      onLayout={[Function]}
      style={
        Object {
          "backgroundColor": "#323232",
          "bottom": 0,
          "opacity": 0,
          "position": "absolute",
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
          "width": "100%",
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hitSlop={undefined}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "opacity": 0,
          }
        }
        testID={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#ffffff",
                "flex": 1,
                "flexWrap": "wrap",
                "marginLeft": 24,
                "marginVertical": 14,
              },
              Object {
                "fontFamily": "Helvetica Neue",
                "marginRight": 24,
              },
            ]
          }
        >
          Snackbar content
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders snackbar with Text as a child 1`] = `
<View
  pointerEvents="box-none"
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    pointerEvents="box-none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      collapsable={undefined}
      onLayout={[Function]}
      style={
        Object {
          "backgroundColor": "#323232",
          "bottom": 0,
          "opacity": 0,
          "position": "absolute",
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
          "width": "100%",
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hitSlop={undefined}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "opacity": 0,
          }
        }
        testID={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#ffffff",
                "flex": 1,
                "flexWrap": "wrap",
                "marginLeft": 24,
                "marginVertical": 14,
              },
              Object {
                "fontFamily": "Helvetica Neue",
                "marginRight": 24,
              },
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            Snackbar content
          </Text>
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders snackbar with action button 1`] = `
<View
  pointerEvents="box-none"
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    pointerEvents="box-none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      collapsable={undefined}
      onLayout={[Function]}
      style={
        Object {
          "backgroundColor": "#323232",
          "bottom": 0,
          "opacity": 0,
          "position": "absolute",
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
          "width": "100%",
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hitSlop={undefined}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "opacity": 0,
          }
        }
        testID={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#ffffff",
                "flex": 1,
                "flexWrap": "wrap",
                "marginLeft": 24,
                "marginVertical": 14,
              },
              Object {
                "fontFamily": "Helvetica Neue",
                "marginRight": 0,
              },
            ]
          }
        >
          Snackbar content
        </Text>
        <View
          style={
            Object {
              "marginHorizontal": 24,
            }
          }
        >
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            hitSlop={undefined}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={undefined}
            testID={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#ff4081",
                }
              }
            >
              UNDO
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders snackbar with content 1`] = `
<View
  pointerEvents="box-none"
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    pointerEvents="box-none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      collapsable={undefined}
      onLayout={[Function]}
      style={
        Object {
          "backgroundColor": "#323232",
          "bottom": 0,
          "opacity": 0,
          "position": "absolute",
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
          "width": "100%",
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hitSlop={undefined}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "opacity": 0,
          }
        }
        testID={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#ffffff",
                "flex": 1,
                "flexWrap": "wrap",
                "marginLeft": 24,
                "marginVertical": 14,
              },
              Object {
                "fontFamily": "Helvetica Neue",
                "marginRight": 24,
              },
            ]
          }
        >
          Snackbar content
        </Text>
      </View>
    </View>
  </View>
</View>
`;
